### DISCUSSION TITLE
missing methods when trying layouts

### DISCUSSION THREAD

**Discussion by tillcarlos** [yippee-fun/phlex#411] [Help] [answered]
Hey community

I'm trying phlex in a new project and I'm also mostly new to phlex. My main problem right now is that there's not much code to copy from. Is there a project set up with best practices?

My current struggle: `undefined method `content_security_policy?'` - that happens in the layout. Not sure if it helps to explain my code much, but I can try

controller:

```ruby
layout false
def weekly
    range = Time.zone.now.all_week
    render Views::Events::Weekly::Index.new() ...
```

The view (shortened)

```ruby
module Views
  module Events
    module Weekly
      class Index < Phlex::HTML
        include ApplicationView
        register_element :turbo_frame
...
        def template
          render Layout.new(title: 'Events Weekly') do
```

application_view - this is modified a bit, the generated one gives me a `undefined local variable or method `csp_meta_tag' ` - I copied some things in, but I guess there are still some missing, causing the error above.

```ruby
# frozen_string_literal: true

require "phlex/rails"

module Views
  module ApplicationView
    include ActionView::RecordIdentifier
    include ActionView::Helpers
    include Propshaft::Helper
    include ActionView::Helpers::AssetTagHelper
    include Phlex::Rails::Layout
    
    include Turbo::StreamsHelper
    include Rails.application.routes.url_helpers
  end
end
```

Any help is appreciated! I'm quite stuck for several hours now, tried to read through everything, but that's as far as I could make it today.

Thanks so much for the great work. Awesome gem - would love to use it!

**Comment by tillcarlos** [comment]
Workaround for now: Just use the ERB layout. Turbo stream can be used with using the helpers like this:

```ruby
helpers.turbo_stream_from "usage_log"
```

Also, I don't see a drawback here doing that. Speed will likely not be affected, and it's a file I almost never touch.

I'll continue with Phlex now, hope I won't write here again (= happy with it) :)

Thanks so much for the great work, guys (h/t  @joeldrapper ), and greetings from Asia.

**Reply by joeldrapper** [reply]
I’m on holiday at the moment, but should be able to look into this in the new year. I’m glad you found a workaround for now!

**Comment by joeldrapper** [comment, marked as answer]
Hey @tillcarlos could you try using the layout pattern described here? https://www.phlex.fun/rails/layouts/

**Reply by tillcarlos** [reply]
I'm sorry to say, I switched the whole thing into ViewComponent.

In the end it was just too hard to wrap my head around how exactly the methods map to HTML. I guess I'm probably too impatient and wanted to get the project done. 

I saw this today: https://github.com/camertron/rux - seems like a cool approach as well. As it compiles it should also be fast.

Thanks for the great work, Joel! We still use Phlex in another project where it solves a huge pain rendering too many views. We are very grateful! You can mark this as resolved if you want.

**Reply by joeldrapper** [reply]
No problem. Glad you found something that worked for you and thanks for the kind words. ❤️
